ui-tests/src/test/java/org/openmrs/reference/page/ManageLocationsPage.java

```diff
@@ -35,23 +48,83 @@ public AddEditLocationPage goToAddLocation() {
         return new AddEditLocationPage(this);
     }
 
-    public void purgeLocation(String name) {
-        clickOn(By.xpath("//tr/td[preceding-sibling::td[contains(text(), '" + name + "')]]/i[@class='icon-trash delete-action']"));
+    public void goToAddNewLocation() {
+        clickOn(ADD_NEW_LOCATION_BUTTON);
+    }
+
+    public void goToEditLocationForm() {
+    	clickOn(EDIT_LOCATION_LINK);
+    }
+    
+    public void enterLocationName(String locationName) {     
+        findElement(LOCATION_NAME_FIELD).clear();
+        findElement(LOCATION_NAME_FIELD).sendKeys(locationName);
+    }
+
+    public void enterLocationDescription(String locationDescription) {
+        findElement(LOCATION_DESCRIPTION_FIELD).clear();
+        findElement(LOCATION_DESCRIPTION_FIELD).sendKeys(locationDescription);
+    }
+
+    public void enterAddress1(String address) {
+        findElement(LOCATION_ADDRESS_FIELD_1).clear();
+        findElement(LOCATION_ADDRESS_FIELD_1).sendKeys(address);
+    }
+    
+    public void enterAddress2(String address) {
+        findElement(LOCATION_ADDRESS_FIELD_2).clear();
+        findElement(LOCATION_ADDRESS_FIELD_2).sendKeys(address);
+    }
+    
+    public void enterState(String stateName) {
+        findElement(STATE_FIELD).clear();
+        findElement(STATE_FIELD).sendKeys(stateName);
+    }
+
+    public void enterCity(String cityName) {
+        findElement(CITY_FIELD).clear();
+        findElement(CITY_FIELD).sendKeys(cityName);
+    }
+    
+    public void enterCountry(String countryName) {
+        findElement(COUNTRY_FIELD).clear();
+        findElement(COUNTRY_FIELD).sendKeys(countryName);
+    }
+    
+    public void enterPostalCode(String postalCode) {
+        findElement(POSTAL_CODE_FIELD).clear();
+        findElement(POSTAL_CODE_FIELD).sendKeys(postalCode);
+    }
+    
+    public void selectParentLocation(String parentLocation) {
+        selectFrom(PARENT_LOCATION_FIELD, parentLocation);
```

Comment:
@kdaud let me put it differently, how do you tell that this is testing the parent location field?

Reply:
@dkayiwa am not sure whether am able to get your point home and am skepticle why am taking a little to get it.

This is what I can tell !
Let me explain what I understand using the UI, Let me share it once more
![Locationx](https://user-images.githubusercontent.com/38831673/132566279-85a5c2fd-a53d-435e-b79f-09b0584e5bf7.png)

Add New Location is a form a system user fills in when creating a custom location within the system. Within  the form there are various fields a user is supposed to fill concerning the location ed Name, Description, Address, Parent Location, etc. Parent Location filed within the form contains a drop down list with location a user can select from.

> how do you tell that this is testing the parent location field?

Not sure whether I am able to get the above question well however,  AFAIK ManageLocationsPage is a page within the system and what the file is containing are page elements a user can interact with when is on the page. So its a test But a method that will be exercised by E2E Location workflow test when creating a location as one of its user story.

@dkayiwa if am reasoning from a different angle, could you once again refrain the question hopefully I will figure it out !

Reply:
What you are explaining is different from what the code is doing.

Reply:
You only need to analyse the contents of the `selectParentLocation()` method, confirm that it is actually testing parent location. Which is not so.

Reply:
Thanks @dkayiwa for clarification ! ooooops I now get the logic, it sounds to me that more code content need to be added to ensure its the parent location that is selected. Initially I thought by declaring a variable(String parentLocation) within selectParentLocation() gives a provision of the test implementation to set the Parent Location used during login instance. Eg if the user is logged using Pharmacy as a location then during the method call, Pharmacy will be set as the string value.

I think renaming the method to setParentLocation() may work since within the test implementation, one has to just set the Location name that is within the drop down list of that field. 
Then when calling the method  within the test implementation we can have some thing like 

manageLocationsPage.setParentLocation("Pharmacy");

@dkayiwa is my idea worth the expectation

Reply:
@kdaud assume you are not the one who created this pull request.
Review the body of the method `selectParentLocation`
Would you approve this pull request and confirm that indeed line 100 is testing parent location?

Reply:
Probably it was an oversight. thanks @dkayiwa  , well done @kdaud

Reply:
> @kdaud assume you are not the one who created this pull request.
> Review the body of the method `selectParentLocation`
> Would you approve this pull request and confirm that indeed line 100 is testing parent location?


Now that have grown through the review, I would not approve the PR simply because technically the method logic does not align with the method name. 

@dkayiwa but now I would approve since the `PR author `has rectified the method name to `setParentLocation()` and it now aligns with the method logic!

Reply:
@kdaud how do you tell that it is testing parent location but not any other field?

Reply:
> @kdaud how do you tell that it is testing parent location but not any other field?

Well, the previous method implementation was providing a room for any other location. As I thought through the better way to implement the method, I think setting the Parent Location within the method would be the best so that the method is called upon without giving provision for a custom location.

@dkayiwa is the idea closely to the expectation ?

Reply:
How do you tell that it is a fixed parent location but not fixed any other field?

Reply:
@dkayiwa have finally gotten a best way to ensure that the `set location `matches with the `parent locatio`n. Have added a logic to cater for ensuring before the location is selected from the drop down list, there is a comparison made first between the set location and the login location. The changes are in the recent last commit with id `c211c38`

Reply:
@kdaud i think you are complicating things way much more than my very simple question. In your `setParentLocation` method, what determines the field you are testing?

Reply:
> In your `setParentLocation` method, what determines the field you are testing? Is it the method name? Or something else?

To my understanding, its the logic of the method that achieves setting Parent Location. However, the set value within the method seParentLocation("setLocation") should match the login session location. 

![Parent](https://user-images.githubusercontent.com/38831673/132976628-6991a120-9fd0-44c9-a7b2-8985cceabdae.png)

@dkayiwa if my reasoning is not closely to the expectation, what is the better way to implement the method  ?

Reply:
What is the id of the field for parent location?

Reply:
> What is the id of the field for parent location?

Its is `selected-location`

Reply:
@kdaud how do you tell that it is what your method is using?

Reply:
> @kdaud how do you tell that it is what your method is using?

Have just written the logic which I think is good enough to achieve the goal though have opted to get the location text via `cssSelector()`.  See code line `100` and `101` in the  updated [commit](https://github.com/openmrs/openmrs-distro-referenceapplication/pull/535/commits/c211c38439357dfe5d2a5902a593e7fa4390c615)

Reply:
@kdaud how do you tell that it is what your method is using?

Reply:
@kdaud `cssSelector` can select any field. So how do you tell that in this method you are selecting the parent location field?

Reply:
> `cssSelector` can select any field. So how do you tell that in this method you are selecting the parent location field?

if (location.matches(findElement(By.cssSelector("span#selected-location")).getText())) {
        	selectFrom(PARENT_LOCATION_FIELD, location);      	
        }

`findElement(By.cssSelector("span#selected-location")).getText()` gets the visible location (text of the location) that was selected during the login session and other field locations are not visible on the page. If the set location is matching the session location visible on the page then the it goes a head to select it from the drop down list else throws an exception.

One challenge with using id in getting the location text is that there are two elements on that page that are using `By.id=selected-location` as their element reference so using id is not a good deal, the test is most likely to some times break while using `By.cssSelector (span#selected-location)` is referencing only one page element(session location).

![Screens](https://user-images.githubusercontent.com/38831673/132992142-a0c2ac1d-37ce-4378-a91d-d19197a48e48.png)

Reply:
What is in `PARENT_LOCATION_FIELD`?

Reply:
> What is in `PARENT_LOCATION_FIELD`?

I am understanding PARENT_LOCATION_FILED from the concept of inheritance . If a class named QualityAssurance extends functionality of OpenMRSCommunity then

OpenmrsCommunity is a parent class of QualityAssurance

                                        `while ` 

QualityAssurance is a child class of OpenMRSCommunity

So PARENT_LOCATION_FIELD am considering it to be the field where the parent location will be selected from.  @kdaud  thinks that the parent location is the location chosen during the user login.

![Parent](https://user-images.githubusercontent.com/38831673/133015679-31381411-f39d-4133-b751-a0c0bdd0ae89.png)

However, @kdaud  might be interpreting parent location field differently which he needs guidance to understand what the exact meaning of PARENT_LOCATION_FIELD .

Reply:
@dkayiwa am tempted to think that may be I did not give the page element (PARENT_LOCATION_FIELD) the right name. If so, what could be the better page element name then ?

Reply:
> @dkayiwa after pausing for a while looking at the reviews on this particular epic, am tempted to think that may be I did not give the right name to the page element (PARENT_LOCATION_FIELD) and its making me to answering reviews in an opposite angle.
> 
> If so, what could be the better name for the page element ?

Hope what am thinking is right, have renamed the page element which I had PARENT_LOCATION_FIELD to PARENT_LOCATION

Reply:
Probably let me change the wording. What value is contained in `PARENT_LOCATION_FIELD`?

Reply:
> What value is contained in `PARENT_LOCATION_FIELD`?

The name of the location. 

![Value](https://user-images.githubusercontent.com/38831673/133045377-6f03e8ef-1fb4-46b1-be26-01163fa83ca6.png)

The location may be Inpatient Ward or Pharmacy or Laboratory, or Isolate Ward or any other location as long as it among the listed location present in the drop down list within the form.

Reply:
Isn't `PARENT_LOCATION_FIELD` a constant?

Reply:
> Isn't `PARENT_LOCATION_FIELD` a constant?

Its a constant and its set within [test.ci.properties](https://github.com/openmrs/openmrs-distro-referenceapplication/blob/b54af82662e7512f5c0cc29ab2b517ab6dd3b983/ui-tests/src/test/resources/org/openmrs/uitestframework/test-ci.properties#L4) file. 

This directly tells me to set it the value directly within the page within the method setParentLocation() and by the method call within the test implementation, the location will be set already.

Reply:
@kdaud does that mean that the `setParentLocation` method is reading it from the test.ci.properties file?

Reply:
> does that mean that the `setParentLocation` method is reading it from the test.ci.properties file?

I think so!

Our CI is using qa-refapp server to spinning our tests against while testing their respective functionality on our system.
 Since the test.ci.property value of the login.location=Pharmacy  then am thinking that setParentLocation() should read it from that file and I just set that specific location within the method.

@dkayiwa if my thinking is not yet closely to the expectation then where should `setParentLocation()`read the parent location from ?

Reply:
@kdaud you are using the `PARENT_LOCATION_FIELD` constant in your `setParentLocation` and you do not know the value contained in this constant???

Reply:
> you do not know the value contained in this constant???

This is fascinating for sure! @kdaud has just figured out the issue after reading carefully the caption above with 3-question marks!

What a bug!
The method was actually doing something different based on the element reference that I had set. 

@dkayiwa you made me dance until I figured out the bug ! Thanks Daniel !

Reply:
> So what do you learn from the longest pull request conversation that we have ever had over the smallest required change?

6 Lessons shouts out loud to me.

1. Analyzing critically the reviewer's comment before I take action. If I had analyzed well the review comment [here](https://github.com/openmrs/openmrs-distro-referenceapplication/pull/535#discussion_r703438019), I would have saved a lot more time than I have consumed over the smallest required change.
2. When reviewing a PR, I must consider helping the PR author to figure out things rather than telling him/her what to do. This can be done through posing precise and short clear questions and in the process, the PR author is able to grow his/her critical thinking!
3. Not to give up on a PR author who may take long to identify what to exact change ks supposed to do as long as he/she keeps responding to the reviewers comment.
4.As a PR author, being upon to the reviewers comments since the reviewers is able to see what I might have not observed during the development.
5. PATIENCE. I can not imagine how long have taken to figure a small change that I could do just within 1 or 2 minute(s), but kept on looking at the reviewer's comment from a different angle and finally was able to identify what was needed.
6. As a reviewer, when I realize the PR author is not failing to identify the required changes, its a good thing to refrain the question perhaps changing wording of the review but remaining on the point.

Thanks @dkayiwa for your patience with me otherwise some one else(reviewer) would have given up ! **Being a mentor requires patience !!**

Reply:
Very well said! 👍